state/immutable-trie/state.go

```diff
@@ -3,122 +3,238 @@ package itrie
 import (
 	"errors"
 	"fmt"
+	"sync"
 
-	lru "github.com/hashicorp/golang-lru"
-
+	"github.com/dogechain-lab/dogechain/helper/hex"
 	"github.com/dogechain-lab/dogechain/state"
 	"github.com/dogechain-lab/dogechain/types"
+	"github.com/hashicorp/go-hclog"
+	"go.uber.org/atomic"
 )
 
-const (
-	codeLruCacheSize         = 8192
-	trieStateLruCacheSize    = 2048
-	accountStateLruCacheSize = 4096
+var (
+	// codePrefix is the code prefix for leveldb
+	codePrefix = []byte("code")
+
+	ErrStateTransactionIsCancel = errors.New("transaction is cancel")
 )
 
-type State struct {
+type StateTransaction interface {
+	Commit() error
+	Cancel()
+}
+
+type State interface {
+	StorageReader
+	StorageWriter
+
+	SetCode(hash types.Hash, code []byte) error
+	GetCode(hash types.Hash) ([]byte, bool)
+
+	NewSnapshot() state.Snapshot
+	NewSnapshotAt(types.Hash) (state.Snapshot, error)
+
+	ExclusiveTransaction(execute func(st StateTransaction))
```

Comment:
It is wired that no error returns when execute failed, and `Rollback()` encapsulated in it.